### PR TITLE
Adjustments to improve GitLab token handling and terminology

### PR DESCRIPTION
**Pull request by Cervator**
## Summary

- Fix BSD sed compatibility (`\?` → explicit `http://`/`https://`) in slug extraction and URL normalization
- Fix nested GitLab group remote naming (second-to-last path segment, not first)
- Add `ws gitlab-auth` command for one-time HTTPS credential setup
- Add multi-token selection (`gp_set_token_for_url`) — longest-prefix match of repo URL against `defaults.gitTokens`
- Add `defaults.upstreamRemote` — overlay-configured default for `ws cr --upstream` with multiple remotes
- Add `@GDD_HOME` placeholder in templates, resolved from `defaults.gddHome` in ecosystem config
- Rename `ws pr` → `ws cr`, `git-pr.sh` → `git-cr.sh`, `.prs/` → `.crs/` for provider-neutral naming
- Auto-source `.env` in main `ws` dispatcher; remove redundant sourcing from sub-scripts
- Enforce AI attribution and `@HUMAN_ACCOUNT` substitution in `git-cr.sh` (matching `git-issue.sh`)
- Docs: GitLab token types, `glab auth login --hostname`, multi-token pattern, getting-started PATH step

## Test plan

- [ ] `ws gitlab-auth` succeeds against self-hosted GitLab URL
- [ ] `ws clone gdd-test` clones via HTTPS without credential prompt
- [ ] `ws push yggdrasil` pushes topic branch to rpraestholm fork
- [ ] `ws cr yggdrasil --upstream "feat: adjust for GitLab" .crs/adjust-for-gitlab.md` opens MR to gni-cis/gdd/yggdrasil

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Environment and drafts
 .env
 .issues/
-.prs/
+.crs/
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply .gitignore:4]: No action — user decision. No backwards-compatibility concern for .prs/ since there is effectively one user.

File: scripts/ws-clone.sh [hunk 2]
```diff
@@ -22,24 +22,44 @@ COMPONENTS_DIR="$ROOT_DIR/components"
...
+        path="/${BASH_REMATCH[1]}"
+    else
+        echo "origin"
+        return
+    fi
+
+    # Strip leading slash, credentials, and .git suffix; split on /
+    path="${path#/}"
+    path="${path%.git}"
+    path="${path%%@*}"   # strip any embedded user:token@ remnant
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws-clone.sh]: Fixed in 4ca5c57 — comment corrected.

File: docs/cr-internals.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+Older glab used `--source-project SLUG` to identify the fork. In this mode,
+glab called the **upstream project's API** — mirroring the GitHub behaviour.
+The upstream reporter token was sufficient.
+
+### glab ≥ 1.65 (--head, current)
+
+`--source-project` was removed and replaced with `--head OWNER/REPO`.
+With this flag, glab calls the **fork project's API** to create the MR —
+not the upstream's API. GitLab's MR creation endpoint lives on the source
+(fork) project.
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply docs/cr-internals.md]: Fixed in 9ab3b43 — removed glab <1.65 section. Docs now describe current --head flow only.

File: docs/dev-setup.md [hunk 4]
```diff
@@ -40,15 +40,15 @@ ws <command> [args...]
 | `ws clone <comp>\|--all` | Clone one or all components |
 | `ws pull [comp]` | Pull latest for all or one component |
 | `ws push [comp] [branch]` | Push via HTTPS (auto-sources .env) |
-| `ws pr <comp> <title> <bodyfile>` | Open a pull request to main |
-| `ws issue <repo> <title> <label> <bodyfile>` | File a GitHub issue |
+| `ws cr <comp> <title> <bodyfile>` | Open a CR (change request/PR/MR) to main |
+| `ws issue <repo> <title> <label> <bodyfile>` | File an issue |
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply docs/dev-setup.md]: Fixed in 9ab3b43 — ws issue <repo> → <comp> [remote].

File: docs/getting-started/index.md [hunk 5]
```diff
@@ -26,11 +26,18 @@ Note that this currently assumes a variety of prerequisites:
...
+   of the longer `bash scripts/ws <command>` form. It also means `ws` always
+   resolves paths relative to the workspace root, so it works correctly no matter
+   which directory you (or an AI agent) run it from:
    ```bash
-   # Add to your ~/.bashrc or shell profile:
-   export PATH="<path-to-yggdrasil>/scripts:$PATH"
+   echo "export PATH=\"$(pwd)/scripts:\$PATH\"" >> ~/.zshrc   # zsh (macOS default)
+   # or:
+   echo "export PATH=\"$(pwd)/scripts:\$PATH\"" >> ~/.bashrc  # bash
    ```
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply docs/getting-started/index.md:39]: Fixed in 9ab3b43 — added blank lines around fenced code block.

File: .env.example [hunk 6]
```diff
@@ -18,13 +18,38 @@ export GH_TOKEN=ghp_xxxxxxxxxxxx
...
 #
-# Required scopes:
-#   api            (full API access — MRs, issues, repo reads)
+# For self-hosted instances, set GITLAB_HOST and register the host with glab once:
+#   source .env && glab auth login --hostname "$GITLAB_HOST" --token "$GITLAB_GDD_GROUP_WRITE_TOKEN"
+#
+# ── Single-group setup (simple) ──────────────────────────────
+# export GITLAB_TOKEN=glpat-xxxxxxxxxxxx        # Group token, Developer role, api scope
+# export GITLAB_USER=your-gitlab-username
+# export GITLAB_HOST=gitlab.example.com         # self-hosted only
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply .env.example:35]: Fixed in 9ab3b43 — self-hosted login example now falls back to GITLAB_TOKEN.

File: docs/cr-internals.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+access level. If your account is Owner of `group/project`, your PAT is effectively
+Owner too — there is no way to issue a PAT that has only Reporter access to a
+group where you have Owner access.
+
+The GitLab-native solution is **Group Access Tokens** and **Project Access Tokens**,
+which are created with an explicit role independent of any user. This is what
+yggdrasil uses on self-hosted/paid tiers (e.g. `GITLAB_UPSTREAM_REPORTER_TOKEN`,
+`GITLAB_FORK_WRITE_TOKEN`). On gitlab.com free tier, these token types are not
+available — use a Personal Access Token instead, and point both env vars at the
+same PAT to maintain the routing pattern.
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply docs/cr-internals.md:139]: Fixed in 9ab3b43 — using actual env var names from .env.example.

File: docs/git-provider-setup.md [hunk 8]
```diff
@@ -90,19 +90,36 @@ After install, open a fresh Git Bash session so the updated `PATH` is picked up.
...
+
+> **gitlab.com free tier:** Group and Project Access Tokens require a paid
+> subscription (Premium+). Use a **Personal Access Token** with `api` scope
+> instead. For the multi-token setup (fork → upstream), you can point multiple
+> env vars at the same PAT — the token routing logic still works, you just
+> don't get the access-level separation that scoped tokens provide.
+
+For Project and Group tokens, set the role to **Developer** (can push branches,
+create MRs and issues). For self-hosted instances, token URLs follow the same
+pattern: `https://<your-host>/<group>/<project>/-/settings/access_tokens`.
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply docs/git-provider-setup.md]: Fixed in 9ab3b43 — added Developer vs Reporter role guidance.

File: scripts/ws-review.sh [hunk 9]
```diff
@@ -313,46 +351,46 @@ if [[ ${#_CANDIDATE_SLUGS[@]} -eq 0 ]]; then
...
-_PEEK_PR=""
+# Peek at the CR number from remaining args to probe which slug has it
+_PEEK_CR=""
 if [[ "${1:-}" == "threads" && "${2:-}" =~ ^[0-9]+$ ]]; then
-    _PEEK_PR="$2"
+    _PEEK_CR="$2"
 elif [[ "${1:-}" =~ ^[0-9]+$ ]]; then
-    _PEEK_PR="$1"
+    _PEEK_CR="$1"
 fi
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws-review.sh:365]: Fixed in 9ab3b43 — added reply to CR probe logic.

**Comment by agent-refr** [comment]
## Deferred review findings → #36

All 11 remaining unresolved threads (7 CodeRabbit + 4 Copilot) have been triaged and collected into #36 as hardening WIBNIFs. None are bugs in tested flows:

- **Script hardening** (4 items): attribution grep scope, sed escaping, ssh URL normalization, unconditional GITLAB_TOKEN export
- **Multi-token GitLab gaps** (3 items): missing token routing in issue filing and review probing, multi-remote collision selector
- **ws gitlab-auth improvements** (4 items): single-token fallback, token in argv, credential helper check, ls-remote probe scope

These affect edge cases in multi-token self-hosted GitLab setups that we cannot fully test on gitlab.com free tier. They will be addressed as those setups become actively used. Resolving threads now with a pointer to the tracking issue.

File: scripts/git-cr.sh [hunk 10]
```diff
@@ -61,6 +58,32 @@ if [[ ! -f "$BODYFILE" ]]; then
...
+fi
+if [[ -z "$_HUMAN_ACCOUNT" ]]; then
+  echo "ERROR: identity.human_account not set in ecosystem config." >&2
+  echo "  Set it in ecosystem.local.yaml (see ecosystem.local.yaml.example)." >&2
+  exit 1
+fi
+if ! grep -q 'AI-assisted' "$BODYFILE"; then
+  echo "ERROR: body file is missing the AI attribution line." >&2
+  echo "  First line must contain: > **AI-assisted change proposal.**" >&2
+  exit 1
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/git-cr.sh:78]: Deferred to #36 — low risk, template controls line 1.

File: scripts/ws [hunk 11]
```diff
@@ -599,6 +604,117 @@ ws_issue() {
...
+                write_token_var=$(KEY="$key" yq '.defaults.gitTokens[strenv(KEY)] // ""' "$eco")
+            fi
+        fi
+    done < <(yq '.defaults.gitTokens | keys | .[]' "$eco" 2>/dev/null)
+
+    if [[ -z "$write_token_var" || "$write_token_var" == "null" ]]; then
+        echo "ERROR: No gitTokens entry found for host '$host' and forkOrg '$fork_org'." >&2
+        echo "  Expected a key like: $host/<group-path>/$fork_org" >&2
+        echo "  Add it to defaults.gitTokens in ecosystem.local.yaml." >&2
+        exit 1
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws:645]: Deferred to #36 — scope creep for this PR.

File: scripts/ws [hunk 12]
```diff
@@ -599,6 +604,117 @@ ws_issue() {
...
+
+    # 2. Store write token in git credential helper for HTTPS clone/push
+    echo "2. Storing write token in git credential helper..."
+    git credential approve <<EOF
+protocol=https
+host=$host
+username=oauth2
+password=$write_token
+EOF
+    echo "   ✓ Stored in $(git config --global credential.helper || echo 'credential store') for $host"
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws:675]: Deferred to #36 — harmless no-op if unconfigured.

File: scripts/ws-review.sh [hunk 13]
```diff
@@ -313,46 +307,46 @@ if [[ ${#_CANDIDATE_SLUGS[@]} -eq 0 ]]; then
...
     if [[ ${#_MATCH_SLUGS[@]} -eq 0 ]]; then
-        echo "ERROR: PR/MR #$_PEEK_PR not found on any remote for '$COMP'." >&2
+        echo "ERROR: CR #$_PEEK_CR not found on any remote for '$COMP'." >&2
         echo "  Tried: ${_CANDIDATE_SLUGS[*]}" >&2
         exit 1
     elif [[ ${#_MATCH_SLUGS[@]} -gt 1 ]]; then
-        echo "ERROR: PR/MR #$_PEEK_PR found on multiple remotes for '$COMP'." >&2
+        echo "ERROR: CR #$_PEEK_CR found on multiple remotes for '$COMP'." >&2
         echo "  Matches: ${_MATCH_SLUGS[*]}" >&2
         echo "  Remove extra remotes or specify which repo to review." >&2
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws-review.sh:397]: Deferred to #36 — needs design for selector mechanism.

File: scripts/git-issue.sh [hunk 14]
```diff
@@ -71,10 +70,12 @@ if ! grep -q 'AI-assisted issue' "$BODYFILE"; then
...
 fi
 
-# Substitute @HUMAN_ACCOUNT placeholder in a temp copy of the body file
+# Substitute @HUMAN_ACCOUNT and @GDD_HOME placeholders in a temp copy of the body file
 RESOLVED_BODY=$(mktemp)
 trap 'rm -f "$RESOLVED_BODY" "$_RESOLVED_ECOSYSTEM" 2>/dev/null' EXIT
-sed "s/@HUMAN_ACCOUNT/@${HUMAN_ACCOUNT}/g" "$BODYFILE" > "$RESOLVED_BODY"
+sed -e "s|@HUMAN_ACCOUNT|@${HUMAN_ACCOUNT}|g" \
+    -e "s|@GDD_HOME|${GDD_HOME}|g" \
+    "$BODYFILE" > "$RESOLVED_BODY"
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/git-issue.sh:78]: Deferred to #36 — values won't contain special chars in practice.

File: scripts/git-provider.sh [hunk 15]
```diff
@@ -107,3 +107,45 @@ gp_detect_and_load() {
...
+
+    if [[ -n "$best_var" && "$best_var" != "null" ]]; then
+        local token_value="${!best_var:-}"
+        if [[ -n "$token_value" ]]; then
+            export GITLAB_TOKEN="$token_value"
+        else
+            echo "WARNING: gitTokens maps '$normalized' to env var '$best_var', but it is not set." >&2
+            echo "  Add it to .env (see docs/git-provider-setup.md)." >&2
+        fi
+    fi
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/git-provider.sh:150]: Deferred to #36 — harmless for GitHub.

File: scripts/ws [hunk 16]
```diff
@@ -599,6 +604,117 @@ ws_issue() {
...
+    done < <(yq '.components[].repo // ""' "$eco" 2>/dev/null)
+
+    if [[ -n "$test_url" ]]; then
+        if git ls-remote "$test_url" HEAD &>/dev/null; then
+            echo "   ✓ git access verified ($test_url)"
+        else
+            echo "   ✗ git ls-remote failed: $test_url" >&2
+            echo "   Token may lack read access to this repo." >&2
+            exit 1
+        fi
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws:692]: Deferred to #36 — edge case in fork verification.

File: scripts/git-provider.sh [hunk 17]
```diff
@@ -107,3 +107,45 @@ gp_detect_and_load() {
...
+
+    local map_count
+    map_count=$(yq '.defaults.gitTokens | length' "$eco" 2>/dev/null || echo 0)
+    [[ "$map_count" -eq 0 ]] && return 0
+
+    # Normalize URL: strip protocol, embedded credentials, .git suffix
+    # Use explicit http/https patterns — \? is GNU sed only, not macOS BSD sed
+    local normalized
+    normalized=$(echo "$url" \
+        | sed 's|^https://[^@]*@||; s|^http://[^@]*@||; s|^https://||; s|^http://||; s|^git@\([^:]*\):|/\1/|; s|^/||; s|\.git$||')
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/git-provider.sh:129]: Deferred to #36 — GDD uses HTTPS everywhere, ssh:// is an edge case.

File: scripts/ws [hunk 18]
```diff
@@ -599,6 +604,117 @@ ws_issue() {
...
+    echo ""
+
+    # 1. Register write token with glab
+    echo "1. Registering write token ($write_token_var) with glab..."
+    if glab auth login --hostname "$host" --token "$write_token" 2>/dev/null; then
+        echo "   ✓ glab authenticated"
+    else
+        echo "   ✗ glab auth login failed" >&2
+        exit 1
+    fi
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws:665]: Deferred to #36 — glab doesn't support stdin token input, accepted trade-off.

File: scripts/git-issue.sh [hunk 19]
```diff
@@ -32,9 +32,6 @@ REPO_ROOT="$SCRIPT_DIR/.."
 # Ensure clearinghouse directory exists
 mkdir -p "$REPO_ROOT/.issues"
 
-# Source .env for provider tokens (GH_TOKEN, GITLAB_TOKEN, etc.)
-_ENV_FILE="$REPO_ROOT/.env"
-[[ -f "$_ENV_FILE" ]] && source "$_ENV_FILE"
 
 # Source provider dispatcher
 # shellcheck source=git-provider.sh
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/git-issue.sh:37]: Deferred to #36 — multi-token gap, works with single PAT.

File: scripts/ws-review.sh [hunk 20]
```diff
@@ -282,9 +320,9 @@ if [[ ! "$COMP" =~ ^[a-z]([a-z0-9-]*[a-z0-9])?(\.[a-z]([a-z0-9-]*[a-z0-9])?)*$ ]
 fi
 
 # Resolve repo slug from component's remotes.
-# Unlike push/issue (which target YOUR fork), review reads PRs that may live
-# on any remote (typically upstream). Strategy: extract the PR number from args,
-# try each remote's slug until the PR is found.
+# Unlike push/issue (which target YOUR fork), review reads CRs that may live
+# on any remote (typically upstream). Strategy: extract the CR number from args,
+# try each remote's slug until the CR is found.
 COMP_DIR="$ROOT_DIR/components/$COMP"
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws-review.sh:329]: Deferred to #36 — multi-token gap in review probes, works with single PAT.

File: scripts/ws-review.sh [hunk 21]
```diff
@@ -265,8 +303,8 @@ fi
 
 # Parse component (first positional arg, always required)
 if [[ $# -lt 1 ]]; then
-    echo "Usage: ws review <comp> threads <pr#> [--status | --resolve <id> | --resolve-all]" >&2
-    echo "       ws review <comp> <pr#> [--reviewer <name>] [--since <time>]" >&2
+    echo "Usage: ws review <comp> threads <cr#> [--status | --resolve <id> | --resolve-all]" >&2
+    echo "       ws review <comp> <cr#> [--reviewer <name>] [--since <time>]" >&2
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws-review.sh:310]: Fixed in 91bfa6d — added reply to header comment and usage output.

File: docs/git-provider-setup.md [hunk 22]
```diff
@@ -111,25 +131,44 @@ Add to `.env` in the yggdrasil root:
...
 
 ### Load and verify
 
+For self-hosted instances, register the hostname with `glab` once after
+setting `GITLAB_HOST`:
+
 ```bash
 source .env
+glab auth login --hostname "$GITLAB_HOST" --token "$GITLAB_TOKEN"
+```
```
Comment (agent-refr) [inline docs/git-provider-setup.md:149]: Fixed in 91bfa6d — uses write-token fallback pattern.

File: .agent/skills/topic-branch-workflow/SKILL.md [hunk 23]
```diff
@@ -19,50 +19,48 @@ Types: `feat`, `fix`, `docs`, `chore`, `test`, `refactor`
 
 Examples: `feat/gh-issue-helper`, `fix/nordri-velero-assert`, `docs/kuttl-gotchas`
 
-## Utility Scripts
+## Workspace CLI Commands
 
-All scripts live in `yggdrasil/scripts/` and auto-source `.env`. Run them from any workspace repo directory.
+Always use `ws` commands instead of raw git operations — they handle staging,
+attribution trailers, auth, and remote selection automatically.
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply .agent/skills/topic-branch-workflow/SKILL.md:25]: The wording is intentional — git checkout/pull are fine, the rule targets git add/commit/push which bypass ws staging and auth. No change needed.
